### Môn Lập trình cho khoa học dữ liệu
#### Giảng viên: Trần Trung Kiên


<h1><center>Đồ án cuối kỳ</center></h1>

MSSV |	Họ và tên | Tên tài khoản github
--- | --- | ---
19120186 |	Đỗ Lê Khánh Đăng | dolekhanhdang
19120412 |	Nguyễn Minh Tú   |  tuak123ax

### Về dữ liệu
Dữ liệu này về lương của những công việc liên quan đến data: data scientist, data engineer, .... <br />
Nguồn: https://www.kaggle.com/nikhilbhathi/data-scientist-salary-us-glassdoor. <br />
Trong mục mô tả của dữ liệu:<br />
- The dataset is scraped from the Glassdoor website using Selenium scrapper. After scrapping, the raw dataset was cleaned and made usable for performing data analysis and modelling. The dataset contains information about the minimum salary, maximum salary, average salary, job description, age of the company in years, etc.
- Dữ liệu được lấy bằng cách scraped từ website của Glassdoor với Selenium scrapper. Sau khi lấy, dữ liệu đã được xử lý phù hợp cho phân tích dữ liệu và mô hình hóa. Tập dữ liệu chứa thông tin về mức lương tối thiểu, tối đa, trung bình, mô tả công việc, tuổi đời công ty,..

## Import thư viện

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Khám phá dữ liệu 

### Đọc dữ liệu từ file csv vào dataframe

In [4]:
salary_df = pd.read_csv('data_cleaned_2021.csv')
salary_df.head()

index                  Job Title              Salary Estimate  \
0      0             Data Scientist   $53K-$91K (Glassdoor est.)   
1      1  Healthcare Data Scientist  $63K-$112K (Glassdoor est.)   
2      2             Data Scientist   $80K-$90K (Glassdoor est.)   
3      3             Data Scientist   $56K-$97K (Glassdoor est.)   
4      4             Data Scientist  $86K-$143K (Glassdoor est.)   

                                     Job Description  Rating  \
0  Data Scientist\r\nLocation: Albuquerque, NM\r\...     3.8   
1  What You Will Do:\r\n\r\nI. General Summary\r\...     3.4   
2  KnowBe4, Inc. is a high growth information sec...     4.8   
3  *Organization and Job ID**\r\nJob ID: 310709\r...     3.8   
4  Data Scientist\r\nAffinity Solutions / Marketi...     2.9   

                                   Company Name         Location  \
0                      Tecolote Research\r\n3.8  Albuquerque, NM   
1  University of Maryland Medical System\r\n3.4    Linthicum, MD   
2                                KnowBe4\r\n4.8   Clearwater, FL   
3                                   PNNL\r\n3.8     Richland, WA   
4                     Affinity Solutions\r\n2.9     New York, NY   

     Headquarters          Size  Founded  ... tensor hadoop tableau bi flink  \
0      Goleta, CA   501 - 1000      1973  ...      0      0       1  1     0   
1   Baltimore, MD       10000+      1984  ...      0      0       0  0     0   
2  Clearwater, FL   501 - 1000      2010  ...      0      0       0  0     0   
3    Richland, WA  1001 - 5000      1965  ...      0      0       0  0     0   
4    New York, NY     51 - 200      1998  ...      0      0       0  0     0   

   mongo  google_an   job_title_sim  seniority_by_title  Degree  
0      0          0  data scientist                  na       M  
1      0          0  data scientist                  na       M  
2      0          0  data scientist                  na       M  
3      0          0  data scientist                  na      na  
4      0          0  data scientist                  na      na  

[5 rows x 42 columns]

### Dữ liệu gồm có bao nhiêu dòng và bao nhiêu cột?

Tính số dòng và số cột sau đó lưu vào 2 biến num_rows và num_cols.

In [6]:
num_rows=salary_df.shape[0]
num_cols=salary_df.shape[1]
print(num_rows)
print(num_cols)

742
42


Dữ liệu có 742 dòng và 42 cột.

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng cho ta biết các thông tin liên quan đến công việc và mức lương của công việc đó. 

Các dòng có ý nghĩa giống nhau.

### Dữ liệu có các dòng bị lặp không?

kiểm tra và lưu kết quả vào biến have_duplicated_rows. Biến này sẽ có giá trị True nếu dữ liệu có các dòng bị lặp và có giá trị False nếu ngược lại.

In [10]:
dup=salary_df.index.duplicated().sum()
if dup==0:
    have_duplicated_rows=False
else:
    have_duplicated_rows=True

In [11]:
have_duplicated_rows

False

Vậy là dữ liệu không có các dòng bị lặp. Ta tiếp tục các công đoạn tiếp theo.

### Mỗi cột có ý nghĩa gì?

Dưới đây là phần mô tả của về các cột trong file "data_cleaned_2021.csv":

index: index

Job Title: Chức danh của công việc, eg. Data scientist, junior data scientist, senior data scientist etc

Salary Estimate: Mức lương và nguồn

Job Description: Cho biết những phẩm chất mà công ty mong muốn và những gì được mong đợi từ chức danh công việc

Rating: đánh giá của công ty

Company Name: tên công ty

Location: Vị trí của công việc

Headquarters: Địa điểm đặt trụ sở chính của công ty

Size: số lượng nhân viên làm việc trong công ty

Founded: Công ty được thành lập vào năm

Type of ownership: Loại hình sở hữu

Industry: Ngành công nghiệp

Sector: ngành

Revenue: doanh thu

Competitors: Đối thủ cạnh tranh

Hourly: hàng giờ

Employer provided: Nhà tuyển dụng cung cấp

Lower Salary: Mức lương dưới

Upper Salary: Mức lương trên

Avg Salary(K): Mức lương trung bình

company_txt: công ty

Job Location: vị trí của công việc

Age: tuổi

Python,spark,aws,excel,sql,sas,keras,pytorch,scikit,tensor,hadoop,tableau,bi,flink,mongo,google_an: các kỹ năng yêu cầu

job_title_sim: chức danh

seniority_by_title: thâm niên

Degree: trình độ

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [13]:
col_dtypes=salary_df.dtypes
col_dtypes

index                   int64
Job Title              object
Salary Estimate        object
Job Description        object
Rating                float64
Company Name           object
Location               object
Headquarters           object
Size                   object
Founded                 int64
Type of ownership      object
Industry               object
Sector                 object
Revenue                object
Competitors            object
Hourly                  int64
Employer provided       int64
Lower Salary            int64
Upper Salary            int64
Avg Salary(K)         float64
company_txt            object
Job Location           object
Age                     int64
Python                  int64
spark                   int64
aws                     int64
excel                   int64
sql                     int64
sas                     int64
keras                   int64
pytorch                 int64
scikit                  int64
tensor                  int64
hadoop    

### Cột có dtype là object nghĩa là sao?

Trong Pandas, kiểu dữ liệu object thường ám chỉ chuỗi, nhưng thật ra kiểu dữ liệu object có thể chứa một đối tượng bất kỳ trong Python (vì thật ra ở bên dưới kiểu dữ liệu object chứa địa chỉ). Nếu một cột trong dataframe có dtype là object thì có thể các phần tử trong cột này sẽ có kiểu dữ liệu khác nhau; để biết được kiểu dữ liệu thật sự của các phần tử trong cột này thì ta phải truy xuất vào từng phần tử và dùng câu lệnh type để xem kiểu dữ liệu.

In [14]:
def open_object_dtype(s):
    dtypes = set()

    # YOUR CODE HERE
    s=s.apply(type)
    dtypes.update(s.unique().tolist())
    
    return dtypes

In [17]:
print(open_object_dtype(salary_df['Job Title']))
print(open_object_dtype(salary_df['Salary Estimate']))
print(open_object_dtype(salary_df['Job Description']))
print(open_object_dtype(salary_df['Company Name']))
print(open_object_dtype(salary_df['Location']))
print(open_object_dtype(salary_df['Headquarters']))
print(open_object_dtype(salary_df['Size']))
print(open_object_dtype(salary_df['Type of ownership']))
print(open_object_dtype(salary_df['Industry']))
print(open_object_dtype(salary_df['Sector']))
print(open_object_dtype(salary_df['Revenue']))
print(open_object_dtype(salary_df['Competitors']))
print(open_object_dtype(salary_df['company_txt']))
print(open_object_dtype(salary_df['Job Location']))
print(open_object_dtype(salary_df['job_title_sim']))
print(open_object_dtype(salary_df['seniority_by_title']))
print(open_object_dtype(salary_df['Degree']))

{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}


Vậy là các cột đều có kiểu dữ liệu là chuỗi, chúng ta không cần xử lý gì thêm.

### Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào?

Các cột dữ liệu dạng numeric là: Rating, Founded, Lower Salary, Upper Salary, Avg Salary(K), Age

Với mỗi cột numeric ta sẽ tính tỉ lệ % giá trị thiếu (từ 0 đến 100), min, max. Sau đó lưu kết quả vào dataframe 'nume_col_df'. Dataframe này có 3 dòng là "missing_ratio", "min", "max"; và có 6 cột là "Rating", "Founded", "Lower Salary", "Upper Salary","Avg Salary(K)","Age".

In [20]:
def XuLy(nume_col_df,columnName):
    list_values=[((salary_df[columnName].isna().sum()/salary_df[columnName].size)*100),
    (salary_df[columnName].min()),
    (salary_df[columnName].max())]
    nume_col_df[columnName]=list_values

In [21]:
cols=['Rating', 'Founded', 'Lower Salary', 'Upper Salary','Avg Salary(K)','Age']
rows=['missing_ratio', 'min', 'max']
nume_col_df=pd.DataFrame(index=rows,columns=cols)
XuLy(nume_col_df,'Rating')
XuLy(nume_col_df,'Founded')
XuLy(nume_col_df,'Lower Salary')
XuLy(nume_col_df,'Upper Salary')
XuLy(nume_col_df,'Avg Salary(K)')
XuLy(nume_col_df,'Age')

In [22]:
nume_col_df

Rating  Founded  Lower Salary  Upper Salary  Avg Salary(K)  \
missing_ratio     0.0      0.0           0.0           0.0            0.0   
min              -1.0     -1.0          15.0          16.0           15.5   
max               5.0   2019.0         202.0         306.0          254.0   

                 Age  
missing_ratio    0.0  
min             -1.0  
max            277.0

min của 3 cột Rating, Founded và Age có vẻ bất thường. Giá trị của 3 cột này không thể là số âm được.

Có vẻ -1 tượng trưng cho missing values trong tập dữ liệu này.

max của cột Age có vẻ là một giá trị nhiễu vì tuổi của con người không thể là 277 được.

### Các câu hỏi ý nghĩa cần trả lời